CHERICap.bsv

```diff
@@ -182,14 +205,65 @@ typeclass CHERICap #( type capT              // type of the CHERICap capability
   function SoftPerms getSoftPerms (capT cap);
   // Set the software permissions
   function capT setSoftPerms (capT cap, SoftPerms softperms);
-  // Get the architectural permissions
+
+  // Legalise hardware permissions
+  function Exact#(capT) legaliseHardPerms(capT cap);
+    let hp = getHardPerms(cap);
+    let oldCap = cap;
+    if (!(hp.permitLoad || hp.permitStore)) begin
+      hp.permitCap = False;
+    end
+
+    if (!(hp.permitCap && hp.permitLoad)) begin
+      hp.permitElevateLevel = False;
+      hp.permitLoadMutable = False;
+    end
+
+    if (!(hp.permitCap)) begin
+      hp.permissionStoreLevel = 0;
+    end
+
+    if (!hp.permitExecute) begin
+      hp.accessSysRegs = False;
+    end
+
+    let some = setHardPerms(cap, hp);
+
+    return Exact {exact: pack(getHardPerms(oldCap)) == pack(getHardPerms(some)), value: some};
+  endfunction
+
+  // Get whether the perms are legal
+  function Bool areLegalHardPerms(capT cap) = legaliseHardPerms(cap).exact;
+
+  // Get all permissions in unlegalised form
+  function Bit #(31) getUnlegalisedPerms (capT cap);
+    let hp = pack(getHardPerms(cap));
+    return zeroExtend ({hp[8:6], 6'b0, getSoftPerms (cap), hp[5:0]});
+  endfunction
+
+  // Get all permissions in legalised form
   function Bit #(31) getPerms (capT cap);
     let hp = pack(getHardPerms(cap));
+    let legalHardPerms = areLegalHardPerms(cap);
+    let mExact = isLegalisedIntMode(cap);
```

Comment:
Oh, that's interesting. Does the spec really say the perms should read zero if the mode bit is inconsistent? I guess it has to be that way for RV32, so RV64 follows suit...

Reply:
Hmm, I don't think it's obvious from the spec, so maybe we need some clarification: could be a Sail bug depending on your perspective.

Reply:
`If the AP-field could not have been produced by ACPERM then all architectural permission bits in
rd are set to 0.`: but the AP-field isnt' defined to include the mode bit.